Some cleanup and miri fixes

Skgland opened the pull request:
- resolve clippy warnings
- run rustfmt
- cleanup leftovers from removed `num` feature
- remove `rust_beta_channel` feature, everything that was gates is stable in the rust-version defined in Cargo.toml
- fix various sources of UB, miri now passes or takes longer than I have patients

Skgland commented:
Marking as Draft as I plan to fix the clippy warnings and want to investigate some of the remaining miri problems

Skgland commented:
@bakaq I think this now resolves all the UB detected with miri, except maybe for some hiding in the test that run very long in miri as I didn't have quite the patients to wait that long.

Skgland commented:
I think this might fix #2391

Skgland commented:
Ah, I think I introduced a memory leak with my fix for arena and stream.

Previously a `Box<TypedAllocSlab<X>>` was allocated and when the `Arena` was dropped `X` was manually dropped in place and a `Box<AllocSlab>` constructed from the `Box<TypedAllocSlab<X>>` was automatically dropped.
The problem with this is that the deallocation of `Box<AllocSlab>` doesn't match the allocation of `Box<TypedAllocSlab<X>>` getting detected as UB.

Some streams could also be dropped early, the `TypedAllocSlab<X>` Payload was dropped in place and the Tag of the `AllocSlab` was changed to Dropped causing the maual drop in place of `X` to be skipped, we couldn't even if we wanted as we have lost the type information.

Now we still allocate a `Box<TypedAllocSlab<X>>` and when the Arena is dropped we reconstruct the `Box<TypedAllocSlab<X>>` to drop it.
This is a problem when a stream is dropped early as we now drop the stream in place, and tag it as Dropped, but now we have lost the type information needed to reconstruct the `Box<TypedAllocSlab<X>>` and we never drop the `Box`.

I am a bit surprised miri's leak detection hasn't caught this, though maybe non of the test that run under miri exercise this edge case.

bakaq commented:
Wow, very impressive. The test suite passing Miri is a good step in verifying the unsafe code in Scryer. This doesn't mean that there isn't UB, just that the tests don't stumble on any. I'm not sure how much the current test suite covers all the relevant unsafe code interleavings (for example, this possible leak edge case that @Skgland described), but this is a very good start.

I think it's a good idea to make some benchmarks of this just to make sure we haven't lost a bunch of performance for some obscure reason, but it's probably fine.

I was going to point out that, just like Wasm, Miri is an interpreter so it makes sense to run it in just one machine in CI, but it seems you already did it like that.

Skgland commented:
> I think it's a good idea to make some benchmarks of this just to make sure we haven't lost a bunch of performance for some obscure reason, but it's probably fine.

A few changes may eliminate optimization potential, but mostly I would assume no change.
Some of the changes (mostly 6386e70) eliminate intermediate allocations and unnecessary clones, so that might improve performance, though I doubt those are in the hot path.
Fixing the pointer alignment UB requires allocating more memory as it now rounds up to the next multiple of the alignment, so that might reduce performance due to allocating more often. 

> 
> I was going to point out that, just like Wasm, Miri is an interpreter so it makes sense to run it in just one machine in CI, but it seems you already did it like that.

Miri is only available in nightly so I just added it to the nightly job.

Skgland commented:
I have removed ArenaHeaderTag::Dropped, wrapping (parts of) the Payload in an Option where it can be dropped early. Also noticed that I was leaking ArenaAllocated IndexPtr and fixed that.

This fixes all issues with this PR I know of. 
The added commits some unsafe by removing manual drop_in_place calls, but add a bunch of unwrap where there are now Option that need to be handled. 

Tests and miri still pass (at least locally, still waiting on CI), so the unwraps should hopefully be unreachable due to checking early whether the stream is missing or not.

Skgland commented:
Looks like some of the miri enables test take a very long time in CI while all run relatively fast locally (under Windows). 
I will check miri tomorrow on my Linus system to attempt to find the culprit test, so that I can disable it for miri and get CI time back down.

Skgland commented:
Other than that I would consider now this ready again.

Skgland commented:
Miri now completes in a resonable time with one additional test disabled🎉

Also I have an Idea for how ArenaHeaderTag::Dropped could work, so I will try that to undo a60c669, as the added `Option`s are not nice.

Skgland commented:
> Also I have an Idea for how ArenaHeaderTag::Dropped could work, so I will try that to undo [a60c669](https://github.com/mthom/scryer-prolog/commit/a60c6695b623c2ef3e00b147dd60404195211720), as the added `Option`s are not nice.

The solution have now took implemented is to add the `ArenaHeaderTag` in `UntypedArenaSlab`, so that we don't lose the type information when the tag is changed to Dropped in the `ArenaHeader` and as such we can correctly deallocate the slab on drop. 
On 64-bit System this increases the `AllocSlab` type by 8-bytes.

Additionally for for the `ArenaAllocated` types where the Payload can be dropped early the Payload is now wrapped in `ManuallyDrop` and manually dropped on drop when not already dropped early, so that we don't double drop the Payload when we drop the Box and it has already been dropped.

This should now be ready again.

Skgland commented:
@mthom soft ping for review, as it has been more than two weeks, would you be able to review this?
Generally probably best reviewed commit by commit, though a couple of things got revised several times.
Given that it has gotten quite large and diverse I can also understand if I should split this up.

mthom commented:
@Skgland Yes, I'm eager to make a review of this starting next week. I spent a 8 days vacation in Newfoundland from which I returned home with Covid, but I'm about fully recovered and should be ready to go Monday.
